accepted/2021/package-validation/package-validation.md

```diff
@@ -0,0 +1,190 @@
+    {
+        // Do network call
+    }
+```
+
+When Finley builds the project it fails with the following error:
+
+> error SYSLIB1234: The method 'DownloadLogs(string)' exists for .NET Standard
+> 2.0 but does not exist for .NET 7.0. This prevents consumers who compiled
+> against .NET Standard 2.0 to run on .NET 7.0.
+
+Finley understands that they shouldn't exclude `DownloadArea(string)` but
+instead just provide an additional `DownloadArea(Utf8String)` method for .NET
+7.0 and changes the code accordingly:
+
+```C#
+#if NET7_0_OR_GREATER
+    public void DownloadLogs(Utf8String area)
+    {
+        // ...
+    }
+#endif
+
+    public void DownloadLogs(string area)
+    {
+        // ...
+    }
+```
+
+### Validation against previous version
+
+Skylar is tasked with adding support for a connection time out to their library.
+The `Connect` method looks like this right now:
+
+```C#
+public static Client Connect(string url)
+{
+    // ...
+}
+```
+
+Since a connection timeout is an advanced configuration setting they reckon they
+can just add an optional parameter:
+
+```C#
+public static Client Connect(string url, TimeSpan? timeout = default)
+{
+    // ...
+}
+```
+
+However, when they rebuild the package they are getting an error:
+
+> error SYSLIB1234: The method 'Connect(string)' exists in the previous version
+> of the NuGet package (AdventureWorks.Client 1.2.3) but no longer exists in the
+> current version being built (1.2.4). This is a breaking change.
+
+Skylar realizes that while this is not a source breaking change, it's a binary
+breaking change. They solve this problem by adding an overload instead:
+
+```C#
+public static Client Connect(string url)
+{
+    return Client(url, Timeout.InfiniteTimeSpan);
+}
+
+public static Client Connect(string url, TimeSpan timeout)
+{
+    // ...
+}
+```
+
+### Making an intentional breaking change
+
+Armani is working on version 0.7 of their library. While they don't want to make
+accidental breaking changes, they do want to make changes based on user feedback.
+
+In version 0.6 they had a `Connect()` method that accepted a timeout as an
+`int`. Users complaint that timeouts in .NET are nowadays generally modelled
+using `TimeSpan` instead. After making the necessary code change, rebuilding the
+library results in an error:
+
+> error SYSLIB1234: The method 'Connect(string, int)' exists in the previous
+> version of the NuGet package (Fabrikam.Client 0.6.3) but no longer exists in
+> the current version being built (0.7.0). This is a breaking change.
+
+Since Armani's library is still an 0.x, which, according to SemVer, doesn't
+promise a stable API, they briefly consider turning off the breaking change
+validation entirely. However, Armani feels that even though it's 0.x they still
+don't want to make accidental breaking change so they decide to just suppress
+this specific instance of the breaking change:
+
+```C#
+[SuppressMessage("Interoperability", "SYSLIB1234")]]
```

Comment:
Do we care about the size footprint that this may add to costumer binaries in metadata? I agree we should support disabling code this way, however should we provide a way to disable errors via an input text file or an MSBuild itemgroup? This could be modeled as tuples:

```
SYSLIB1234, Connect(System.String, System.TimeSpan)
```

or 
```xml
<PackageValidation Include="SYSLIB1234" Member="Connect(System.String, System.TimeSpan)" />
```

Reply:
Do we have precedent for designating types/members in MSBuild like this? I wonder whether it would be fragile ..

Reply:
I'm not sure; but I would imaging adding these attributes for all the APICompat errors we baseline in dotnet/runtime for example which is a pretty large codebase would not be a good option because of the metadata bloat those would cause.

Reply:
cc: @ericstj for opinions/ideas as well.

Reply:
Another option would be #pragma's that could be read by the tool, either through symbol->source mapping or read by an analyzer and shuffled over to the API compat tool through a file on the side, perhaps even the PDB (same could be done through Conditional attributes).  Those could mitigate size impact on assembly.  I think authoring in source is significantly better than msbuild project file, which is better than random xml/text file.

In this particular case I see the attribute is applied to the new method.  That could be problematic.  When analyzing metadata we wouldn't necessarily see the removal of one overload and the addition of a new one as the same thing.  In the case that many overloads were added, which one should get the suppression for the removed overload?  When members are completely removed, where do you put the suppression?  API Compat solved this with a crude method: suppress by the full log string.

Another concern: we'd want to make sure folks aren't unnecessarily suppressing warnings.  Folks might also want to have many runs of the tool (eg: once against previous major version, once against previous minor).  Where would they put the different suppressions?  Also, if they were only running against previous version, would we ask them to remove all the suppressions once they passed the single version that introduced the breaking change?

Reply:
> Another option would be #pragma's that could be read by the tool, either through symbol->source mapping or read by an analyzer and shuffled over to the API compat tool through a file on the side, perhaps even the PDB (same could be done through Conditional attributes). Those could mitigate size impact on assembly. I think authoring in source is significantly better than msbuild project file, which is better than random xml/text file.

Wouldn't that mean that if we wanted to run against the previos version of the package which had it's own APICompat suppressions we would need to run against the symbols package if the PDBs are not part of the package that we downloaded?

Also we would need to come up with our own #pragma format that included RID/TFM/SemVer if you wanted to suppress for a specifc major version but not for others? Or if you wanted to have a suppression for just one RID but not for the others, etc?

Reply:
> Wouldn't that mean that if we wanted to run against the previos version of the package which had it's own APICompat suppressions we would need to run against the symbols package if the PDBs are not part of the package that we downloaded?

Understanding if these suppressions need to travel with the binary itself is an important distinction.  That's not clear to me today: in all scenarios I've seen, right-hand-side is being built.  If we want to support otherwise then I agree we should have a reliable persistent storage location.

> Also we would need to come up with our own #pragma format that included RID/TFM/SemVer if you wanted to suppress for a specifc major version but not for others? Or if you wanted to have a suppression for just one RID but not for the others, etc?

Source already has conditional syntax (conditions in project file, `#ifdef` in source).  For applying suppressions to right-hand-side I think we should use that.  Constraining the suppression based on what's on the left-hand-side could be interesting.  I think we've mentioned a few different interesting variables: version, TFM, RID.  What do we think these constraints might help us with?  

For version my thought was that it would allow us to "permit" old suppressions to remain if the version on the left-hand-side was newer than that being analyzed.  This way we wouldn't force people to delete suppressions the release after they made the breaking change (they still could if they wanted).

I'm not sure if it's so interesting for TFM/RID: if you are going to permit a break you typically are OK with that break from any source on the left-hand-side.  To handle the "unused" case here, we could look at all compat runs in a batch and only raise an unused warning if none of them required the suppression (and the version rule didn't apply).

Reply:
Yeah I can't think of a scenario where left would need to carry suppressions over, but maybe there is one. 

Also, something that might be interesting would be that we are running APICompat after pack (I don't know if the opt-in against a previous version would be after build or after pack, I would guess after pack so that we can actually do package validation againt it as well not just API Compat?). Any way when running after pack it would be trickier to get the PDBs as that happens in the outer build where there is no RID/TFM context, I know we could figure out the PDBs paths with some targets and stuff, but how ideal would that be? 

What I'm trying to say is how ideal is depending on PDBs as well?

Reply:
Just brainstorming existing files that might already flow through the process where we could stash metadata.  PDBs definitely have some downsides.  Assembly metadata is, of course, easier. Was just reading our [docs on suppress message](https://docs.microsoft.com/en-us/visualstudio/code-quality/in-source-suppression-overview?view=vs-2019) and found this that addresses metadata-bloat:
> The SuppressMessageAttribute attribute is a conditional attribute, which is included in the IL metadata of your managed code assembly, only if the CODE_ANALYSIS compilation symbol is defined at compile time.

> *You should not use in-source suppressions on release builds, to prevent shipping the in-source suppression metadata accidentally. Additionally, because of the processing cost of in-source suppression, the performance of your application can be degraded.*

Maybe if we use it that way it makes it OK?  Folks would need to pack debug builds to have it work.

> Also something to consider is how are we going to baseline errors for Package Validation itself, like package dependencies errors and those kind of errors? Or I guess those shouldn't be allowed to be disabled.

We should also have a path for suppressions that isn't exclusively assembly metadata.  NoWarn, but also more granular suppressions.  We'd need to think about all the encoding we'd want to do on those.

Reply:
> Maybe if we use it that way it makes it OK? Folks would need to pack debug builds to have it work.

That would make me a little nervous though, as not a lot of PR/CI validation builds run dotnet pack on Debug builds, I don't know how common that practice is? And I would guess release pipelines always pack on Release configuration, so is a behavior where people can only suppress messages on one configuration but not in the other? That could loose some fidelity as our tooling would only be running on Debug builds that run pack.

Also, if we are going to do suppressions on source as our first class mode to suppress errors I think we should provide a fixer in the near future...